Fixed i2c dependency loop

swift-tk opened the pull request:
This was discovered when trying to add select GPIO to emulator Kconfig.

Try add "select GPIO" to zephyr/drivers/espi/Kconfig.espi_emul and use
west build -p always -b native_sim tests/drivers/espi/
Error log: [log.txt](https://github.com/zephyrproject-rtos/zephyr/files/15388645/log.txt)

swift-tk committed:
boards: native: fixed i2c dependency loop

This was discovered when trying to add select GPIO to emulator Kconfig.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
tests: subsys: fix emul test compile

Add CONFIG_EMUL=y to the test .conf

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
tests: sensor: fix compile

Add CONFIG_EMUL=y to prj.conf

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk committed:
samples: subsys: fix sensing/simple compile

Add CONFIG_EMUL=y to native_sim.conf

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

aescolar reviewed (commented):
Are you sure that you want to do select GPIO?
https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-recommendations

swift-tk commented:
> Are you sure that you want to do select GPIO? https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-recommendations

I saw that, but I think it is ok to have higher level module select a lower level one. Plus @henrikbrixandersen said I could use select GPIO when we chat on discord.

aescolar commented:
The problem with selects is that we end up with loops all around the place.

And the thing with this change is that it changes a default that has been there for many years. So it will cause some breaks for users.

Due to that I would lean towards ESPI_EMUL just depending on GPIO, and setting the CONFIG_GPIO in the configuration files where needed for its tests.

------

Sidetrack:
> Nevertheless, EMUL should not depend on I2C.

That "depends" is not a "real" thing. In the sense that it is "OR'ed" with the no-dependency of the original kconfig option definition and becomes "true" always. It appears in menuconfig because of the "if I2C" just above, which is also modulating the `default y` to be effectively `default y if I2C && BOARD_NATIVE_SIM`.

swift-tk commented:
> The problem with selects is that we end up with loops all around the place.

I agree, but I believe I'm using it correctly as stated in the doc below.

> An exemption are busses like I2C and SPI, and following the same thought process things like MFD as well. Drivers on these busses should use select to allow the automatic activation of the necessary bus drivers when devices on the bus are enabled in the devicetree.

> And the thing with this change is that it changes a default that has been there for many years. So it will cause some breaks for users.
> Due to that I would lean towards ESPI_EMUL just depending on GPIO, and setting the CONFIG_GPIO in the configuration files where needed for its tests.
I think it is included in ece74b7b7f4 by Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>, Wed Jul 05 2023 (11 months ago)

In my usecase, GPIO would be needed by the driver not the test actually. That would adding CONFIG_GPIO=y to everything project that uses the driver, which is against what is said on the doc.

IMO, this is obviously wrong and a misuse according to the doc. It is a select EMUL essential, thats why EMUL depends on I2C.  

```
if I2C

config EMUL
	default y

endif # I2C
```

IMHO, we should be afraid of touching the old code when needed for integrity.

swift-tk commented:
> And the thing with this change is that it changes a default that has been there for many years. So it will cause some breaks for users.

That may not be true , as I have seen in tree, most of people have avoided select EMUL and added CONFIG_EMUL=y to project even under sensors and i2c.

swift-tk commented:
Another way that might work is to add config EMUL default y without the if I2C, this shouldn't break the samples and test I modified, but I don't know if there are any side effects.

swift-tk committed:
samples: lvgl: fix accelerometer_chart compile

Add CONFIG_EMUL=y to prj.conf.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk commented:
Should have everything fixed now. @aescolar How shall we proceed?

aescolar commented:
>  @aescolar How shall we proceed?

@swift-tk We can go with this, but please add a point in the migration guide telling users this setting does not default to y anymore for this targets.

swift-tk committed:
doc: add a note in 3.7 migration guide

CONFIG_EMUL is no longer default with CONFIG_I2C as it causes I2C
dependency loop.

Signed-off-by: Swift Tian <swift.tian@ambiq.com>

swift-tk commented:
@aescolar @henrikbrixandersen @teburd 
Everything should be ready now. Please help approve again and merge. Thanks.

aescolar commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -35,6 +35,10 @@ Boards
   All symbols are renamed as follows: ``SOC_IT81202BX``, ``SOC_IT81202CX``, ``SOC_IT81302BX``,
   ``SOC_IT81302CX``, ``SOC_IT82002AW``, ``SOC_IT82202AX``, ``SOC_IT82302AX``.
   And, rename the ``SOC_SERIES_ITE_IT8XXX2`` to ``SOC_SERIES_IT8XXX2``. (:github:`71680`)
+* native: The :kconfig:option:`CONFIG_EMUL` is no longer enabled by default with
+  :kconfig:option:`CONFIG_I2C` in :ref:`native_posix<native_posix>` and
+  :ref:`native_sim<native_sim>` Kconfig.defconfig. Out-of-tree users should add
+  :kconfig:option:`CONFIG_EMUL`=y in their project level. (:github:`73067`)
```
```suggestion
* For native_sim/posix: :kconfig:option:`CONFIG_EMUL` is no longer enabled by default when
  :kconfig:option:`CONFIG_I2C` is set. Users who need this setting enabled should set it in their project config file. (:github:`73067`)
```

swift-tk replied:
done
